mep-0001-ray-future-based-shuffle.md

```diff
@@ -0,0 +1,167 @@
+# Motivation
+Shuffle is the core operation for large-scale data processing. The data is partitioned according to the partition key in each record through the ALL-TO-ALL connection, and sent from all mappers to all reducers. Due to the nature of its ALL-TO_ALL connections, shuffle often becomes the bottleneck of the performance, stability and scalability of the entire system. This MEP introduces a pull-based shuffle implementation using Ray future which serves following goals:
+- Improve Mars-on-Ray shuffle scalability by reducing shuffle meta
+- Improve scheduling performance by using Ray C++ scheduler
+- Support shuffle failover by lineage reconstruction
+- Setup the foundation for push-based shuffle using Ray future
+ 
+## Design and Architecture
+When the Mars subtask has upstream and downstream dependencies, the downstream subtask's dependence on the output of the upstream subtask is determined by the data key. The input data required for subtask execution is a context dict object from data key to data.
+ 
+For non-shuffle operators, the data key is the chunk key; but for shuffle operators, the data key is a tuple of (chunk key, reducer index).
+ 
+Therefore, in order to build a Shuffle DAG based on Ray future, either all the data keys output by each mapper subtask are known in advance, or the data keys can be mocked when executing subtask.
+ 
+### Building shuffle DAG by reducer index
+Actually, Mars' ChunkGraph has information about the upstream and downstream chunk dependencies. After converting to Ray DAG, this part of the information still exists. The reason for relying on data keys is that Mars replaced `ShuffleProxy` with `FetchShuffle`. The implementation of `Fetch` and `FetchShuffle` is an abstraction of P2P, which is inconsistent with the semantics of Ray DAG actually. Therefore, when using Ray Backend, a new set of Shuffle operators can be introduced to replace `FetchShuffle` and rely on Ray to automatically resolve subtask dependencies.
+ 
+#### Operand Changes
+- Add `n_reducers` and `reducer_ordinal` to all `MapReduceOperand` operands
+- Introduce a new group of `FetchShuffle` operands which only records `n_mappers` and `n_reducers` to reduce meta overhead:
+```python
+class FetchShuffleType(Enum):
+   FETCH_SHUFFLE_BY_INDEX = 0
+   FETCH_SHUFFLE_BY_DATA_KEY = 1
+ 
+class FetchShuffleByIndex(Operand):
+   _op_type_ = opcodes.FETCH_SHUFFLE_BY_INDEX
+   # number of all upstream mappers
+   n_mappers = Int32Field("n_mappers")
+   # number of all downstream reducers
+   n_reducers = Int32Field("n_reducers")
+ 
+class FetchShuffle(Operand):
+   _op_type_ = opcodes.FETCH_SHUFFLE_BY_DATA_KEY
+   source_keys = ListField("source_keys", FieldTypes.string)
+   source_idxes = ListField("source_idxes", FieldTypes.tuple(FieldTypes.uint64))
+   source_mappers = ListField("source_mappers", FieldTypes.uint16)
```

Comment:
Use `FetchShuffleByIndex` to replace `FetchShuffle`? If so, `FetchShuffle` can be removed.

Reply:
`FetchShuffle` will still be used by mars shuffle

Reply:
Can we just record the n_mappers and n_reducers in `FetchShuffle`? and I think for Mars shuffle we can remove the `source_keys` stuff as well, but for compatibility we can mix them together, I think we don't need a new opcode.

Reply:
If mars can remove `source_keys ` stuff too, add `n_mappers` and `n_reducers` would be better.